sylt-compiler/src/typechecker.rs

```diff
@@ -1804,8 +1796,23 @@ impl TypeChecker {
                     }
                 }
 
-                (Type::Function(a_args, a_ret), Type::Function(b_args, b_ret)) => {
+                (
+                    Type::Function(a_args, a_ret, a_purity),
+                    Type::Function(b_args, b_ret, b_purity),
+                ) => {
                     // TODO: Make sure there is one place this is checked.
+                    match (a_purity, b_purity) {
+                            (Purity::Undefined, _) |
```

Comment:
Maybe call it `Unknown`? Maybe? I don't know?

Reply:
Or just have `Some(Pure)`, `Some(Impure)` and `None`. That would allow us to use the same enum to model situations when the potential purity must be known.

Reply:
```
pub enum Purity {
	Pure,
	Impure,
}
```
sounds like a `bool`. `struct Pure(bool);`? this becomes¹
```
match (a_purity, b_purity) {
	(None, _) |
	(_, None) |
	(Some(a), Some(b)) if a == b => (),
	// ...
}
```


but i guess this becomes a bit verbose. i'd be happy with `enum Purity { Pure, Impure }` as well but wanted to raise `struct Pure(bool)` just in case. thoughts welcome.

---
¹ although, this could be accomplished with an enum too if we derive Eq